scripts/add-license.js

```diff
@@ -0,0 +1,127 @@
+governing permissions and limitations under the License.
+*/
+const fs = require("fs");
+const path = require("path");
+const stagedGitFiles = require("staged-git-files");
+const Handlebars = require("handlebars");
+
+const PROJECT_ROOT = path.resolve(__dirname, "../");
+const SOURCE_TEMPLATE = "source-header.handlebars";
+
+const GIT_DELETED = "Deleted";
+const SOURCE_FILE_EXTENSIONS = ["js", "ts", "cjs", "mjs"];
+const IGNORE_PATTERNS = [
+  /\/sandbox\//gi,
+  /\/scripts\//gi,
+  /launch.+\.js/gi,
+  /at\.js/gi,
+  /\.min\.js/gi,
+  /AppMeasurement/gi
+];
+
+const walk = async (dir, matchesFilter) => {
+  let files = fs.readdirSync(dir);
+  files = await Promise.all(
+    files
+      .filter(file => matchesFilter(file, dir))
+      .map(async file => {
+        const filePath = path.join(dir, file);
+        const stats = fs.statSync(filePath);
+        if (stats.isDirectory()) {
+          return walk(filePath, matchesFilter);
+        }
+        return filePath;
+      })
+  );
+
+  return files.reduce((all, folderContents) => all.concat(folderContents), []);
+};
+
+const getStagedGitFiles = async () => {
+  return (await stagedGitFiles())
+    .filter(detail => {
+      const parts = detail.filename.split(".");
+      return (
+        detail.status !== GIT_DELETED &&
+        parts.length > 1 &&
+        SOURCE_FILE_EXTENSIONS.indexOf(parts[1]) > -1
+      );
+    })
+    .map(detail => path.join(PROJECT_ROOT, detail.filename));
+};
+
+const getAllSourceFiles = async () => {
+  const IGNORED = ["node_modules", ".git", "dist"];
+  return walk(PROJECT_ROOT, (file, dir) => {
+    const filePath = path.join(dir, file);
+    const stats = fs.statSync(filePath);
+
+    for (let i = 0; i < IGNORED.length; i += 1) {
+      if (filePath.includes(IGNORED[i])) {
+        return false;
+      }
+    }
+
+    if (stats.isDirectory()) {
+      return true;
+    }
+
+    if (stats.isFile()) {
+      for (let i = 0; i < SOURCE_FILE_EXTENSIONS.length; i += 1) {
+        if (file.endsWith(SOURCE_FILE_EXTENSIONS[i])) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  });
+};
+
+const run = async () => {
+  const stagedOnly = typeof process.env.STAGED_ONLY !== "undefined";
+
+  const template = fs.readFileSync(
+    path.resolve(__dirname, SOURCE_TEMPLATE),
+    "utf-8"
+  );
+
+  const renderTemplate = Handlebars.compile(template);
+
+  const templateText = renderTemplate({
+    year: new Date().getFullYear()
+  });
+
+  const sourceFiles = stagedOnly
+    ? await getStagedGitFiles()
+    : await getAllSourceFiles();
+
+  sourceFiles
+    .filter(file => {
+      for (let i = 0; i < IGNORE_PATTERNS.length; i += 1) {
+        if (file.match(IGNORE_PATTERNS[i])) {
+          return false;
+        }
+      }
+      return true;
+    })
+    .forEach(file => {
+      const contents = fs.readFileSync(path.resolve(file), "utf-8");
+      if (templateText.slice(0, 2) !== contents.slice(0, 2)) {
```

Comment:
no that code is comparing the first 2 characters of the file with the first two of the license header.  Those two characters are always `/*`.  In other words, the script assumes that if the source file begins with a comment, it contains a license already.  It may not get every case, but will satisfy 99.9%